Issue/9728 posts remote preview

develric opened the pull request:
### Fixes #9728 
This PR adds support for Remote Preview of blog Posts vs the currently available local preview.
To this goal we need that the Post the user wants to preview is previously synchronized with the server.
In this PR we are mainly interested in the following categories of Posts:

1. Draft: this exists already on the server actually
2. Local Draft: a draft that exists on only the user's device (both entirely or partially)
3. New Post
4. Published and Scheduled Post with local changes

All of these will be saved locally and synched with the server if needed. To avoid the unintended publication of a published post for the category 4 the remote auto save will be used instead.
(**NOTE**: this PR needs the FluxC modifications for remote auto save available in the _master-remote-auto-save_ branch, see 18205856d74bb04747484182a056741275d97efb).

### User Flow
1. User asks for preview from Posts List or from Editor
2. In case of draft with no modification the remote preview is directly opened
3. In case of new post or local draft, the user is presented with the following progress dialog while the post is uploading and then the remote preview is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59987617-1de08100-963b-11e9-954c-672173654c1f.png">
</p>
4. In case of published or scheduled post with local changes, the user is presented with the following dialog while the post is remote auto save and then the remote preview is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59988484-7dd72780-963b-11e9-84be-b89aff91764d.png">
</p>
Special cases:

1. If there is no network, the following view is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483312-37b24180-9c95-11e9-981d-48c75233a477.png">
</p>
2. If the connected site is a self-hosted NOT Jetpack powered site, the following is shown (see  https://href.li/?https://github.com/wordpress-mobile/WordPress-iOS/pull/11944#pullrequestreview-251304944 .
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483354-59abc400-9c95-11e9-8e7d-bdcecb75e599.png">
</p>

3. If an error occurs during preview operation the following snackbar is shown
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483599-6aa90500-9c96-11e9-9551-b0d4d23d9fea.png">
</p>


### Background Implementation notes
- In the Editor and in the Posts List, a state machine was added to manage the state of the preview and the presented progress dialog. 
- An EventBus sticky event `PostPreviewingInEditor` was defined in `EditPostActivity` to signal when the preview is ongoing from the editor. This way we can make some event filtering in the Posts List.
- Since all the upload operations are currently expecting an `OnPostUploaded` event, when we are remote auto saving, we chain on remote auto save response receiving a OnPostUploaded event to conclude the uploading transactions. **NOTE:** in case of errors during uploading, the UI currently present to the user the possibility to Retry the upload. For published posts that fail during remote auto saving for preview, as first solution (but in my opinion it can make sense) we skip the error condition (and an eventual retry that can publish the local changes) by not relaunching the error condition in the OnPostUploaded. In all other cases the user is as usual triggered to Retry upload.

### To test:
1. To be able to ask for preview in the posts list, you need to create local changes; this can be easily done in the editor, and published posts simply keep them in the posts list if you don't explicitly publish them. For drafts you can exit from the editor while in offline mode.
2. To get a failure during upload for preview I used the following: in a self hosted with Jetpack site, disable the jetpack plugin and ask for a preview. You will get an error during preview operations. (I'm sure there are better solutions, but it worked for me 🙂).

This said:

**Test Case 1**
1. Ask for preview connecting to a self-hosted not jetpack site and observe you get the Preview Unavailable view.

**Test Case 2**
1. Connected to a WPcom site, put device offline and ask for preview. Observe you get the no network webview.

**Test Case 3**
1. From the editor, ask for preview on draft with no local changes
2. No progress shown and you get the remote preview

**Test Case 4**
1. Ask for preview on draft with local changes or new post
2. Observe the Saving progress and then get the remote preview

**Test Case 5**
1. Ask for preview on published post with local changes
2. Observe the Generating Preview progress and then get the remote preview
3. repeat previous steps for scheduled post with local changes

**Test Case 6**
1. Create a test scenario for failure during preview creation (for example with the disable of the jetpack plugin as said above)
2. Ask for preview for a draft with local changes and observe you get the following error snackbar 
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483698-e30fc600-9c96-11e9-90ec-75aafe0aa561.png">
</p>

and you are allowed to Retry upload
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59990448-7ddb2580-9643-11e9-8709-bd1549ddfe8e.png">
</p>

**Test Case 7**
1. Create a test scenario for failure during preview creation (for example with the disable of the jetpack plugin as said above)
2. Ask for preview for a published post with local changes and observe you get the same error snackbar as before but you are not presented to Retry
<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/59990585-f5a95000-9643-11e9-80de-61c30cbd0384.png">
</p>
3. repeat previous steps for scheduled post with local changes

### Open Items
These are open items that will be faced with dedicated Issue/PR and probably need further discussion. I opened related Issues for them and will follow up based on how things progress here.
- [x] Scheduled posts should be assimilated to published posts and treated the same; currently FluxC allows remote auto save of only Published posts. So need to modify this behavior in FluxC. This current PR implementation tries to manage them as draft. (see [FluxC#1297](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1297) )
- [ ] Trashed posts management in preview. (see #10106)
- [x] Menu items and actions available in the WPWebViewActivity when used to preview posts in each of the above status. (see #10107)
- [ ] There is a [known issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/11684) for Jetpack powered sites and remote auto save; I think there is already API side work triggered from that so will follow it for resolution.
- [ ] Remote preview operation and dialog cancellation: see #10141. 
 
**Update release notes:**

- [x] Since there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -566,30 +567,84 @@ public static void viewHistoryDetailForResult(Activity activity, Revision revisi
+            SiteModel site,
+            PostModel post,
+            RemotePreviewType remotePreviewType) {
+        if (
```
Nitpicking: do you mind splitting this in 2 tests for readability / style.

develric replied:
👍 done here 2ff1e77e6da180b8e89f08d725ea9a18117f5f39

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -566,30 +567,84 @@ public static void viewHistoryDetailForResult(Activity activity, Revision revisi
 
         // always add the preview parameter to avoid bumping stats when viewing posts
-        String url = UrlUtils.appendUrlParameter(post.getLink(), "preview", "true");
+        String url =
```
Do you mind moving this to a Utils method like `getPreviewUrlForPost`: 1. would be a bit easier to read and 2. it might be useful elsewhere.

develric replied:
Yep, it makes sense. Done here cc3d05fc36d6f7e89aec9326032a929d1d2ca8fa

maxme commented on RELEASE-NOTES.txt:
```diff
@@ -1,6 +1,7 @@
 12.8
 -----
-Add percentage bar to following Stats cards - Posts and Pages, Authors, Tags and Categories
+* Added Remote Preview support for Blog Posts
```
IMHO "Add Remote Preview support for Drafts and Published posts with local changes" would be better.

develric replied:
Considering some work to do with the open item related tickets, it definitely better describes what this PR brings. I just added also Scheduled and moved to 12.9. Changed here 4992a250e24b3658bf67e65282ba1ede9cd7a302.

maxme replied:
👍 We'll probably merge this in a feature branch (until we fix other known issues and have the corresponding fluxc branch merged in develop), so I guess we'll bump version number again later.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -419,6 +436,46 @@ class PostListMainViewModel @Inject constructor(
+        /* We need only transitions, so... */
+        if (_previewState.value == newState) return
+
+        AppLog.d(AppLog.T.POSTS, "****** POSTLIST PREVIEW SM - from [${_previewState.value}] to [$newState]")
```
I'd probably remove that log all together, not sure if you used it for debugging or if you intend to track bugs from our logs using that.

If you want to keep it, can we remove the formatting and caps here, we don't do this in other log entries.

develric replied:
It remained there from a debug session; but thinking at it again if you don't mind I would prefer to leave it (changing the formatting) to trace eventual bugs in future. Changed here 02f8676ecca8bb8ae8e3b8e70a3b150214c7ce9c. Let me know if it's fine for you.

maxme commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1455,6 +1455,15 @@
+    <!-- Post Remote Preview -->
+    <string name="post_preview_saving_draft">Saving&#8230;</string>
+    <string name="post_preview_remote_auto_saving_post">Generating Preview&#8230;</string>
+    <string name="remote_preview_operation_error">Error during remote preview preparation.\nPlease try again.</string>
```
I think we can do better on the error wording, do you mind updating it to: `Error while trying to save post before preview` or something similar. I checked on iOS, they show the preview empty screen when an error happens. TBH I think it's better to show a Snackbar in the case of an error.

develric replied:
I was also not happy with the wording actually 😅 ; your suggestion sounds good to me. Changed here 5951e1a694a041d01e1e20ba726b2e97c910bcbc. Also agree that snackbar logic can be better in case of a generic error like this.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1506,20 +1655,58 @@ private void showDialogError(@StringRes int resIdMessage, @StringRes int resIdPo
+    }
+
+    private void updatePostLoadingAndDialogState(PostLoadingState postLoadingState, @Nullable PostModel post) {
+        /* We need only transitions, so... */
```
Super nitpicking here and bellow: can you use `//` for single line comments (I actually thought it was in our checkstyle config).

develric replied:
👍 changed here d3958f7e6bd55072e65c999c3f0ebbf3e303a486

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -3993,61 +4207,99 @@ public void onPostChanged(OnPostChanged event) {
-                mPost = post;
-                mIsNewPost = false;
-                invalidateOptionsMenu();
+            if (!(mPostLoadingState == PostLoadingState.UPLOADING_FOR_PREVIEW)
```
This method is very hard to read and quite hard to understand. I know `EditPostActivity` is not a model of good practices, but can you split this into smaller methods and add some comments to explain what's going on?
Some comments like: "When we get a post update originating from the preview feature: open the internal browse for this post".

develric replied:
You are absolutely right, I have to admit I did not get it while coding but it is not clear to read and understand. I tried to improve it here 42d13f43b94884e1457fbbe4aee99a757a470133. Hope it is better.

maxme replied:
Looks better thanks!

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1506,20 +1655,58 @@ private void showDialogError(@StringRes int resIdMessage, @StringRes int resIdPo
+
+        AppLog.d(
+                AppLog.T.POSTS,
+                "****** EDITOR PREVIEW SM - from [" + mPostLoadingState + "] to [" + postLoadingState + "]"
```
Same comment as above about the log.

develric replied:
For same reason as above I would keep it reformatting. Done in same commit as above d3958f7e6bd55072e65c999c3f0ebbf3e303a486

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/RemotePreviewLogicHelper.kt:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    interface RemotePreviewHelperFunctions {
+        fun isNewPost() = false
```
We should remove `isNewPost()` from the preview logic. `post.isLocalDraft` will always be true when `isNewPost` is true. It seems that `isNewPost` was introduced to show some dialog/info in the editor and do some analytics tracking. Unfortunately other devs later used it for publishing logic and it was probably a mistake.

develric replied:
Understand the rational and thanks for sharing the background, it seems obvious I didn't get it fully 😊. Makes sense to remove isNewPost and canDiscard as you suggest in this next comment. Changes here e8e97532281aa2adce1528bce409a30f303dff00

maxme replied:
Changes looks good in e8e9753

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -1349,6 +1415,70 @@ private boolean handleBackPressed() {
+
+            @Override
+            public boolean canDiscard() {
+                return isDiscardable();
```
I think we want to use `PostUtils.isPublishable` instead of `isDiscardable` here. I'd also rename it if we change that. I guess we can remove `canDiscard` from the helper completely (and use the PostUtils directly).

`isDiscardable` check if the post is "freshly created" but that should not impact what kind of save we do before showing the preview.

develric replied:
Yep, removed canDiscard from the helper and used PostUtilityWrapper for testability/mocking purposes. Please see again e8e97532281aa2adce1528bce409a30f303dff00.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:
```diff
@@ -23,7 +39,49 @@ sealed class PostListAction {
+                    PostListRemotePreviewState.REMOTE_AUTO_SAVING_FOR_PREVIEW,
+                    PostInfoType.PostNoInfo
+            )
+            if (!UploadService.isPostUploadingOrQueued(post)) {
```
I'd consider logging a message in the else statement. The method is called `startUploading` but it sometimes doesn't actually start any upload. Logging a message could help us debugging possible issues. Wdyt?

develric replied:
👍 Done here 4a8bd66598bbd7b7179e3b37fdaf62dc02fafbd5.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -276,6 +287,11 @@ class PostsListActivity : AppCompatActivity(),
 
+    override fun onPause() {
+        super.onPause()
+        progressDialog?.dismiss()
```
The dialog is re-shown when the activity gets re-created, but I think it'll simply disappear if the activity is just paused and resumed. Wdyt?

Note: I'm not 100% sure about this, but I didn't have time to give it a try. I thought it'd be better mentioning it.

develric replied:
Good catch by just having a look at the code 😅! I regret to admit this was an oversight from my side! It was part of a snippet I missed to remove while exploring a path for dialog dismiss; finally just added this general precaution adding a preventive dismiss here in ProgressDialogHelper.kt

``` kt
             is ProgressDialogUiState.VisibleProgressDialog -> {
                dialog?.dismiss()
```

(maybe redundant but not harmful and localized) and stayed with what was already in the EditPostActivity (or at least that was the intention 🙂) keeping similar also the PostsListActivity dialog management. Blame on me and really thanks for having pointed it out 🙇‍♂️. 

While reviewing this point I realized that in PostsListActivity I did not managed the Progress Dialog state in cases when the activity is destroied while in background (same as when Don't keep activities developer option is on), so I added this management that is already present in EditPostActivity.

Changes here f93f82ee76a1cbdec795279dd93150cf39316c6f.

malinajirka replied:
>I regret to admit this was an oversight from my side! It was part of a snippet I missed to remove while exploring a path for dialog dismiss; 

No worries at all, dialogs can be pretty tricky. However, I think the "fix" has broken something else. If we don't dismiss the dialog when the activity is about to be destroyed, I believe it'll leak the view. 

>While reviewing this point I realized that in PostsListActivity I did not managed the Progress Dialog state in cases when the activity is destroied while in background (same as when Don't keep activities developer option is on), so I added this management that is already present in EditPostActivity.

Hmm I'm wondering if this is what we want. The ViewModel will survive (and retain the state) in most cases and if it doesn't survive its highly probable that the app was in the background for a longer period of time. I think in an ideal scenario we'd cancel the request when the ViewModel's "onCleared" method was called (however, I'm a bit afraid it's not a simple task with the current implementation of PostStore).

I also noticed the dialog is not cancelable - it seems a bit inconvenient for the user. Imagine they unintentionally click on "Preview" and they are on a poor connection - they end up with a blocking dialog for 30+ seconds. 

I think these two issues might actually result in an infinite dialog in some cases. I'm not sure if it can really happen, but my gut tells me it's risky. Imagine both the activity and the VM is destroyed -> we unsubscribe from the dispatcher in `PostListEventListener.onDestroy` -> we miss the `OnPostChanged` event -> the user re-opens the app -> we restore the progress dialog's state -> we will wait indefinitely for an event which will never arrive (since it already arrive)..... This is an issue which you could have not seen coming, so even if it can happen don't worry about it. 

Both my concerns are related to canceling the ongoing request. I can't think of an easy and clean solution right now. I also don't want to make this PR more complicated. I'd consider creating an issue and fixing it later when we have some time to think about it. 
Wdyt? cc @maxme

develric replied:
Created ticket #10141 to trace this and update list of open items at the end of this PR description.

develric commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListActivity.kt:
```diff
@@ -227,6 +231,13 @@ class PostsListActivity : AppCompatActivity(),
             it?.let { snackBarHolder -> showSnackBar(snackBarHolder) }
         })
+        viewModel.previewState.observe(this, Observer {
+            progressDialog = ProgressDialogHelper.updateProgressDialogState(
```
Yep it makes sense and it is more in line with other cases in my code already. I have to admit I previously heard about @Reusable scope but did not used it before. So it was a good TIL moment time 😊 and performance-wise seems good in cases like this. Thanks for sharing 🙇‍♂️. Changes here d3958f7e6bd55072e65c999c3f0ebbf3e303a486

malinajirka replied:
Thanks! Tbh I started using @Reusable just a few months ago :P.

develric commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -229,6 +238,8 @@ class PostListMainViewModel @Inject constructor(
                 authorFilterSelection = authorFilterSelection,
                 authorFilterItems = getAuthorFilterItems(authorFilterSelection, accountStore.account?.avatarUrl)
         )
+        _previewState.value = _previewState.value ?: PostListRemotePreviewState.NONE
```
😊…what was that, some sort of mind reading? 😅...you actually got exactly the path I went through. I needed to reassign the live data for the reason you highlighted. Initially I was tempted to align the start method to the approach with flag I noticed in other VMs, but preferred to stay with just that simple modification of the code I was introducing in case I was missing some rationale behind that 😊.

malinajirka replied:
I'm really glad to hear you were thinking about it and that you got to the same conclusion. The decision to go with the simple modification seems reasonable to me in this case ;)!

develric commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt:
```diff
@@ -419,6 +436,46 @@ class PostListMainViewModel @Inject constructor(
+                    }
+                }
+            }
+            else -> { /* nothing to do*/ }
```
I think it is a more explicit and more "under control" approach and I like it. In general I guess could be a must when using 'when' (unwanted play on words 😊) as an expression in kotlin so you fully take advantage of kotlin error level checks (if not used as an expression AFAIU you get a warning I think). I would propose the following:
- in EditPostActivity.java I would list all the possible enum values to be explicit and add a default with throw exception.
- In PostListMainViewModel.kt maybe seems odd to list all possible cases and add an else (that AS reports as redundant) to throw an exception. So for now I (hopefully 😅) reached the goal managing the when as an expression body function (actually it could be an interesting pattern to decide to use when only as an, eventually dummy, expression in this or maybe better way 🤔). Also, rethinking to it, here for the nested when it actually seems to be more appropriate to use an if instead. Changes here d3958f7e6bd55072e65c999c3f0ebbf3e303a486 in `managePostLoadingStateTransitions` and `managePreviewStateTransitions`.

malinajirka replied:
>in EditPostActivity.java I would list all the possible enum values to be explicit and add a default with throw exception.

>In PostListMainViewModel.kt maybe seems odd to list all possible cases and add an else (that AS reports as redundant) to throw an exception.

> So for now I (hopefully 😅) reached the goal managing the when as an expression body function (actually it could be an interesting pattern to decide to use when only as an, eventually dummy, expression in this or maybe better way 🤔). 

Sorry, I should have been more explicit. I believe you can simply omit the `default/else` case if you list all the possible values. `Lint/ktlint` will abort the build if a `case` is missing (works on enums) -> which is even better than throwing an exception as you are informed about the issue before runtime. Wdyt?

develric replied:
ok, I see 😊. Removed default branch in EditPostActivity.java. For PostListMainViewModel.kt since in any case I refactored to have a function, I staid with the expression body function but removed the comment since now is useless to underline that. Changes here d2dc6e2de721107aa3b5cfa9f1519cd1469b578f

develric commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:
```diff
@@ -32,7 +90,27 @@ fun handlePostListAction(activity: FragmentActivity, action: PostListAction) {
+                )
+
+                if (opResult == RemotePreviewLogicHelper.PreviewLogicOperationResult.OPENING_PREVIEW) {
+                    // TODO: evaluate if this case can really happen in the Posts List.
```
AFAIU (a confirmation from your side is more tha welcome of course 😊) currently the PREVIEW button is available in a post item in the Posts List only if isLocalDraft or isLocallyChanged (https://github.com/develric/WordPress-Android/blob/010931d67d965d2406f8bbf40b1829f91107d7b5/WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt#L286). So in this PR, Draft, Published and Scheduled posts will not get here. Trashed posts can reach here currently, but I think they will be managed differently soon based on this #10106.
So i would keep for now the if check here but changing the TODO comment for example as in here a05aba815ad943d5430d38074ef54ae1baace40c. wdyt?
Pinging also @maxme on this since he triggered the above ticket.

malinajirka replied:
The updated comment makes it more clear and actionable even for someone who doesn't know the code so well -> 👍. I'd also consider updating the description of #10106 with a note to remove this if statement, so we don't forget about it. Wdyt?

maxme replied:
I agree we should keep the TODO here - note that we'll probably merge this PR to its own feature branch once we solve some known issues. Good point, @malinajirka we should comment on #10106 to make sure we don't forget it.

develric replied:
Added note about this in #10106 description 👍.

develric commented:
Hi @maxme and @malinajirka 😊.

I should have followed up to all your comments and request (in case I missed something just let me know 🙇‍♂️).

I tried to keep commits separated per comment/request as I could. 

Thanks to your effort to let me have immediately the FluxC modified version (thanks again for your incredible team work 😎) I included the modification for scheduled posts that now should behave as published for preview respect (updated the test cases in the PR description to mark those that must be repeated also for scheduled)

About the WPWebViewActivity:
- used it to report the no network condition instead of using the snackbar (also updated the PR description for this)
- Customized the webview menu so to have:

**standard menu** when using it in the standard way
<p align="center">
<kbd>
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483115-6ed42300-9c94-11e9-9278-70a5470a7d71.png">
</kbd>
</p>

**preview menu** when using it for remote preview (see issue #10107 )
<p align="center">
<kbd>
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483135-7dbad580-9c94-11e9-9ca4-fda411e532ef.png">
</kbd>
</p>

**no menu** when using directly the actionable empty view (like for preview not available)
<p align="center">
<kbd>
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483150-975c1d00-9c94-11e9-97ef-997fecc1bc3e.png">
</kbd>
</p>


- Added Post Title in the Action Bar in case of direct usages of Actionable Empty View (for now Preview Unavailable and No Network Error)
<p align="center">
<kbd>
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60483189-b8bd0900-9c94-11e9-8407-e09723f214b2.png">
</kbd>
</p>

PS: during this review session I noticed that talking about posts without local changes, when you VIEW (not PREVIEW) them and from the WPWebViewActivity you open the external browser, only pulished posts can be actually seen; for other categories you get something like

<p align="center">
<kbd>
  <img width="300" src="https://user-images.githubusercontent.com/47797566/60482647-b22d9200-9c92-11e9-9493-86bc2a839e7f.png">
</kbd>
</p>

I tried a version of the app before my modifications and AFAIU this behaviour was already present so should not have been introduced by the new code. Don’t know if it can be expected but wanted to mention (I can eventually open a ticket for this).


Should be all from my side, I will wait for your feedbacks 😊.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/posts/RemotePreviewLogicHelper.kt:
```diff
@@ -0,0 +1,106 @@
+        return PreviewLogicOperationResult.GENERATING_PREVIEW
+    }
+
+    private fun shouldSave(post: PostModel): Boolean {
```
Sorry I haven't done that comment in my first review pass, but can we rename `shoulSave()` method to `shouldUpload()` just to make sure we don't mix it up with "local" save.

develric replied:
Yep, you are right, it makes sense. Changed here 4839b9a04e50c0b4f9683ea634fa7712e687959f

maxme commented:
> I tried a version of the app before my modifications and AFAIU this behaviour was already present so should not have been introduced by the new code. Don’t know if it can be expected but wanted to mention (I can eventually open a ticket for this).

That's right, just tried it on Trashed post, I thought in all these cases we were showing a "local preview" instead of that. I don't think it's necessary to open a new tickets considering we have #10106 and #9729

develric commented:
Hi @maxme  and @malinajirka , waiting to decide how to manage this PR (probably merge in a feature branch) as agreed I updated for tracking purposes the list of open items at the end of the PR description; also made a follow up to some comments still marked as not resolved; namely these four:
-	[default/else in switch/when](https://github.com/wordpress-mobile/WordPress-Android/pull/10085#discussion_r300503188)
-	[update ticket with note to remove TODO](https://github.com/wordpress-mobile/WordPress-Android/pull/10085#discussion_r300503746)
-	[preview operation and dialog cancellation](https://github.com/wordpress-mobile/WordPress-Android/pull/10085#discussion_r300503909)
-	[rename shouldSave to shouldUpload](https://github.com/wordpress-mobile/WordPress-Android/pull/10085#discussion_r300504052)

Hope it’s fine 😊.

maxme commented:
Note: I updated the target branch to `issue/9728-posts-remote-preview` (avoid conflicts with develop and let us push more commits if we need it).
